src/main/java/com/flipkart/ranger/healthservice/ServiceHealthAggregator.java

```diff
@@ -0,0 +1,171 @@
+        this.inlineHealthMonitorList = Lists.newArrayList();
+        this.running = new AtomicBoolean(false);
+    }
+
+    /**
+     * this method can be used to add a {@link IsolatedHealthMonitor} which will later be
+     * scheduled at regular intervals and monitored to generate and maintain an aggregated health of the service
+     * the scheduling will happen in an isolated thread
+     *
+     * @param monitor any extension of the {@link IsolatedHealthMonitor}
+     */
+    @Override
+    public void addIsolatedMonitor(IsolatedHealthMonitor monitor) {
+        if (running.get()) {
+            /* cant add monitors when the Aggregator is already running */
+            throw new UnsupportedOperationException("Cannot add a monitor when Aggregator is running");
+        }
+        isolatedHealthMonitorList.add(monitor);
+    }
+
+    /**
+     * this method can be used to add a {@link Monitor}
+     * this monitor will not be scheduled in a separate isolated thread,
+     * but instead its execution will happen inline, in a single thread, along with other inline monitors
+     *
+     * @param monitor an implementation of line {@link Monitor<HealthcheckStatus>}
+     */
+    @Override
+    public void addInlineMonitor(Monitor<HealthcheckStatus> monitor) {
+        if (running.get()) {
+            /* cant add monitors when the Aggregator is already running */
+            throw new UnsupportedOperationException("Cannot add a monitor when Aggregator is running");
+        }
+        inlineHealthMonitorList.add(monitor);
+    }
+
+    /**
+     * start monitoring all registered monitors
+     * (triggers a scheduled execution of all registered monitors) and saves their futures for later reference)
+     */
+    @Override
+    public void start() {
+        if (running.get()) {
+            /* in case the monitor is already running */
+            return;
+        }
+        ScheduledExecutorService scheduledExecutorService = Executors.newScheduledThreadPool(isolatedHealthMonitorList.size());
+        scheduledFutureList = Lists.newArrayListWithCapacity(isolatedHealthMonitorList.size());
+        for (IsolatedHealthMonitor isolatedHealthMonitor : isolatedHealthMonitorList) {
+            final ScheduledFuture<?> scheduledFuture =
+                    scheduledExecutorService.scheduleAtFixedRate(
+                            isolatedHealthMonitor,
+                            isolatedHealthMonitor.getRunInterval().getInitialDelay(),
+                            isolatedHealthMonitor.getRunInterval().getTimeInterval(),
+                            isolatedHealthMonitor.getRunInterval().getTimeUnit());
+            scheduledFutureList.add(scheduledFuture);
+        }
+        running.set(true);
+    }
+
+    /**
+     * stop all running monitors
+     */
+    @Override
+    public void stop() {
+        for (ScheduledFuture<?> scheduledFuture : scheduledFutureList) {
+            scheduledFuture.cancel(true);
```

Comment:
noted, stop must be a synchronized process. Also there needs to be a double check on if the state is running or not. I'll add in those changes